File: examples/low-level-api/merkle_key.rs
// Copyright 2020-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! An example that revokes a key and shows how verification fails as a consequence.
//!
//! cargo run --example merkle_key

use identity::core::Timestamp;
use identity::credential::Credential;
use identity::crypto::merkle_key::Sha256;
use identity::crypto::merkle_tree::Proof;
use identity::crypto::KeyCollection;
use identity::crypto::PrivateKey;
use identity::crypto::PublicKey;
use identity::did::MethodScope;
use identity::iota::CredentialValidationOptions;
use identity::iota::CredentialValidator;
use identity::iota::Receipt;
use identity::iota::Resolver;
use identity::iota_core::IotaDID;
use identity::iota_core::IotaVerificationMethod;
use identity::prelude::*;
use rand::rngs::OsRng;
use rand::Rng;

mod common;
mod create_did;

#[tokio::main]
async fn main() -> Result<()> {
  // Create a client instance to send messages to the Tangle.
  let client: Client = Client::new().await?;

  // Create a signed DID Document/KeyPair for the credential issuer (see create_did.rs).
  let (mut issuer_doc, issuer_key, issuer_receipt): (IotaDocument, KeyPair, Receipt) = create_did::run().await?;

  // Create a signed DID Document/KeyPair for the credential subject (see create_did.rs).
  let (subject_doc, _, _): (IotaDocument, KeyPair, Receipt) = create_did::run().await?;

  // Generate a Merkle Key Collection Verification Method with 8 keys (Must be a power of 2)
  let keys: KeyCollection = KeyCollection::new_ed25519(8)?;
  let method_did: IotaDID = issuer_doc.id().clone();
  let method = IotaVerificationMethod::new_merkle_key::<Sha256>(method_did, &keys, "merkle-key")?;

  // Add to the DID Document as a general-purpose verification method
  issuer_doc.insert_method(method, MethodScope::VerificationMethod)?;
  issuer_doc.metadata.previous_message_id = *issuer_receipt.message_id();
  issuer_doc.metadata.updated = Timestamp::now_utc();
  issuer_doc.sign_self(issuer_key.private(), issuer_doc.default_signing_method()?.id().clone())?;

  // Publish the Identity to the IOTA Network and log the results.
  // This may take a few seconds to complete proof-of-work.
  let receipt: Receipt = client.publish_document(&issuer_doc).await?;
  println!("Publish Receipt > {:#?}", receipt);

  // Create an unsigned Credential with claims about `subject` specified by `issuer`.
  let mut credential: Credential = common::issue_degree(&issuer_doc, &subject_doc)?;

  // Select a random key from the collection
  let index: usize = OsRng.gen_range(0..keys.len());

  let public: &PublicKey = keys.public(index).unwrap();
  let private: &PrivateKey = keys.private(index).unwrap();

  // Generate an inclusion proof for the selected key
  let proof: Proof<Sha256> = keys.merkle_proof(index).unwrap();

  // Sign the Credential with the issuers private key
  issuer_doc
    .signer(private)
    .method("merkle-key")
    .merkle_key((public, &proof))
    .sign(&mut credential)?;

  println!("Credential JSON > {:#}", credential);

  // Check the verifiable credential is valid
  let resolver: Resolver = Resolver::new().await?;
  let resolved_issuer_doc = resolver.resolve_credential_issuer(&credential).await?;
  CredentialValidator::validate(
    &credential,
    &resolved_issuer_doc,
    &CredentialValidationOptions::default(),
    identity::iota::FailFast::FirstError,
  )?;

  println!("Credential successfully validated!");

  // The Issuer would like to revoke the credential (and therefore revokes key at `index`)
  issuer_doc
    .resolve_method_mut("merkle-key", None)?
    .revoke_merkle_key(index as u32)?;
  issuer_doc.metadata.previous_message_id = *receipt.message_id();
  issuer_doc.metadata.updated = Timestamp::now_utc();
  issuer_doc.sign_self(issuer_key.private(), issuer_doc.default_signing_method()?.id().clone())?;

  let receipt: Receipt = client.publish_document(&issuer_doc).await?;

  println!("Publish Receipt > {:#?}", receipt);

  // Check that verifiable credential is revoked

  let resolved_issuer_doc = resolver.resolve_credential_issuer(&credential).await?;
  let result: Result<()> = CredentialValidator::validate(
    &credential,
    &resolved_issuer_doc,
    &CredentialValidationOptions::default(),
    identity::iota::FailFast::FirstError,
  )
  .map_err(Into::into);

  assert!(result.is_err());

  println!("Credential successfully revoked!");

  Ok(())
}

File: bindings/wasm/examples/src/merkle_key.js
// Copyright 2020-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    Client,
    Credential,
    CredentialValidationOptions,
    CredentialValidator,
    Digest,
    FailFast,
    KeyCollection,
    KeyType,
    MethodScope,
    Resolver,
    SignatureOptions,
    Timestamp,
    VerificationMethod
} from '@iota/identity-wasm';
import {createIdentity} from './create_did';

/**
 This example shows how to sign/revoke verifiable credentials on scale.
 Instead of revoking the entire verification method, a single key can be revoked from a MerkleKeyCollection.
 This MerkleKeyCollection can be created as a collection of a power of 2 amount of keys.
 Every key should be used once by the issuer for signing a verifiable credential.
 When the verifiable credential must be revoked, the issuer revokes the index of the revoked key.

 @param {{network: Network, explorer: ExplorerUrl}} clientConfig
 **/
async function merkleKey(clientConfig) {
    // Create a client instance to publish messages to the configured Tangle network.
    const client = await Client.fromConfig({
        network: clientConfig.network
    });

    // Creates new identities (See "create_did" example)
    const alice = await createIdentity(clientConfig);
    const issuer = await createIdentity(clientConfig);

    // Add a Merkle Key Collection Verification Method with 8 keys (Must be a power of 2)
    const keys = new KeyCollection(KeyType.Ed25519, 8);
    const method = VerificationMethod.newMerkleKey(Digest.Sha256, issuer.doc.id(), keys, "key-collection")

    // Add to the DID Document as a general-purpose verification method
    issuer.doc.insertMethod(method, MethodScope.VerificationMethod());
    issuer.doc.setMetadataPreviousMessageId(issuer.receipt.messageId());
    issuer.doc.setMetadataUpdated(Timestamp.nowUTC());
    issuer.doc.signSelf(issuer.key, issuer.doc.defaultSigningMethod().id());

    // Publish the Identity to the IOTA Network and log the results.
    // This may take a few seconds to complete proof-of-work.
    const receipt = await client.publishDocument(issuer.doc);
    console.log(`Identity Update: ${clientConfig.explorer.messageUrl(receipt.messageId())}`);

    // Prepare a credential subject indicating the degree earned by Alice
    let credentialSubject = {
        id: alice.doc.id().toString(),
        name: "Alice",
        degreeName: "Bachelor of Science and Arts",
        degreeType: "BachelorDegree",
        GPA: "4.0"
    };

    // Create an unsigned `UniversityDegree` credential for Alice
    const unsignedVc = Credential.extend({
        id: "https://example.edu/credentials/3732",
        type: "UniversityDegreeCredential",
        issuer: issuer.doc.id().toString(),
        credentialSubject,
    });

    // Sign the credential with Issuer's Merkle Key Collection method, with key index 0
    const signedVc = issuer.doc.signCredential(unsignedVc, {
        method: method.id().toString(),
        public: keys.public(0),
        private: keys.private(0),
        proof: keys.merkleProof(Digest.Sha256, 0)
    }, SignatureOptions.default());

    // Check the verifiable credential is valid
    const resolver = await Resolver
        .builder()
        .client(client)
        .build();
    const resolvedIssuerDoc = await resolver.resolveCredentialIssuer(signedVc);
    CredentialValidator.validate(
        signedVc,
        resolvedIssuerDoc,
        CredentialValidationOptions.default(),
        FailFast.FirstError
    );
    console.log(`Credential successfully validated!"`);

    // The Issuer would like to revoke the credential (and therefore revokes key 0).
    issuer.doc.revokeMerkleKey(method.id().toString(), 0);
    issuer.doc.setMetadataPreviousMessageId(receipt.messageId());
    issuer.doc.SetMetadataUpdated(Timestamp.nowUTC());
    issuer.doc.signSelf(issuer.key, issuer.doc.defaultSigningMethod().id());
    const nextReceipt = await client.publishDocument(issuer.doc);
    console.log(`Identity Update: ${clientConfig.explorer.messageUrl(nextReceipt.messageId())}`);

    // Check the verifiable credential is revoked
    let vc_revoked = false;
    try {
        // Resolve the issuer's updated DID Document to ensure the key was revoked successfully.
        const updatedResolvedIssuerDoc = await resolver.resolveCredentialIssuer(signedVc);
        CredentialValidator.validate(
            signedVc,
            updatedResolvedIssuerDoc,
            CredentialValidationOptions.default(),
            FailFast.FirstError
        );
    } catch (exception) {
        console.log(`${exception.message}`)
        vc_revoked = true;
    }

    if (!vc_revoked) throw new Error("VC not revoked");
    console.log(`Credential successfully revoked!`);
}

export {merkleKey};
